[Win] Enable LTO for release builds

iangrunert opened the pull request:
#### 3431171a0e69487a4f5536c157b3b5522617c010
<pre>
[Win] Enable LTO for release builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=281502">https://bugs.webkit.org/show_bug.cgi?id=281502</a>

Reviewed by Fujii Hironori.

Changes to set-webkit-configuration to handle path separator differences
Changes to WebKitCompilerFlags.cmake to pass LTO params to clang-cl

* Source/cmake/WebKitCompilerFlags.cmake:
* Tools/Scripts/set-webkit-configuration:
* Tools/Scripts/webkitdirs.pm:
(determineLTOMode):

Canonical link: <a href="https://commits.webkit.org/285241@main">https://commits.webkit.org/285241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/746290b6c53ca0d7de5c256761568aa23a108b45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71845 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76003 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23050 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73960 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22870 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56726 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15222 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61900 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19371 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21400 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64974 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65082 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77680 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71100 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16079 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18923 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64671 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61923 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64451 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6266 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92880 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11044 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47058 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20465 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48129 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49413 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47871 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii reviewed (commented):
Does this change enable LTO by default? How much slower to build?

fujii commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -253,12 +253,18 @@ if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" AND NOT "${LOWERCASE_CMAKE_HOST_SY
     set(CMAKE_EXE_LINKER_FLAGS "-flto=${LTO_MODE} ${CMAKE_EXE_LINKER_FLAGS}")
     set(CMAKE_SHARED_LINKER_FLAGS "-flto=${LTO_MODE} ${CMAKE_SHARED_LINKER_FLAGS}")
     set(CMAKE_MODULE_LINKER_FLAGS "-flto=${LTO_MODE} ${CMAKE_MODULE_LINKER_FLAGS}")
+elseif (LTO_MODE AND COMPILER_IS_CLANG AND MSVC AND NOT DEVELOPER_MODE)
```
What's the reason you add `NOT DEVELOPER_MODE`?

iangrunert replied:
There's a couple of link errors for WebCore with LTO in developer mode:

```
lld-link: error: undefined symbol: __declspec(dllimport) currentStackPointer
>>> referenced by S:\WebKit\WebKitBuild\Debug\WebCore\DerivedSources\unified-sources\UnifiedSource-68aea4ac-21.cpp
>>>               Source\WebCore\CMakeFiles\WebCore.dir\__\__\WebCore\DerivedSources\unified-sources\UnifiedSource-68aea4ac-21.cpp.obj

lld-link: error: undefined symbol: private: static class WebCore::BreakLines::LineBreakTable const WebCore::BreakLines::lineBreakTable
>>> referenced by S:\WebKit\WebKitBuild\Debug\WebCore\DerivedSources\unified-sources\UnifiedSource-207b877e-6.cpp
>>>               Source\WebCore\CMakeFiles\WebCore.dir\__\__\WebCore\DerivedSources\unified-sources\UnifiedSource-207b877e-6.cpp.obj
>>> referenced by S:\WebKit\WebKitBuild\Debug\WebCore\DerivedSources\unified-sources\UnifiedSource-043dd90b-15.cpp
>>>               Source\WebCore\CMakeFiles\WebCore.dir\__\__\WebCore\DerivedSources\unified-sources\UnifiedSource-043dd90b-15.cpp.obj
>>> referenced by S:\WebKit\WebKitBuild\Debug\WebCore\DerivedSources\unified-sources\UnifiedSource-043dd90b-18.cpp
>>>               Source\WebCore\CMakeFiles\WebCore.dir\__\__\WebCore\DerivedSources\unified-sources\UnifiedSource-043dd90b-18.cpp.obj
ninja: build stopped: subcommand failed.
```

The first is something across a DLL boundary but the second I'm not sure.

iangrunert committed:
[Win] Enable LTO for release builds
https://bugs.webkit.org/show_bug.cgi?id=281502

Reviewed by Fujii Hironori.

Changes to set-webkit-configuration to handle path separator differences
Changes to WebKitCompilerFlags.cmake to pass LTO params to clang-cl

* Source/cmake/WebKitCompilerFlags.cmake:
* Tools/Scripts/set-webkit-configuration:
* Tools/Scripts/webkitdirs.pm:
(determineLTOMode):

Canonical link: https://commits.webkit.org/285241@main
